### PR TITLE
Add Aspire.Hosting.Testing to facilitate integration testing

### PR DESCRIPTION
**Pull request by ReubenBond**
Addresses #1704

This PR adds two types to aid in integration testing of Aspire AppHost projects.

* `DistributedApplicationTestingBuilder<TEntryPoint>`

This type allows you to write tests following the builder pattern. For example, if this is your app host's Program.cs:

```csharp
var builder = DistributedApplication.CreateBuilder(args);
builder.AddRedis("redis1");
builder.AddProject<Projects.TestingAppHost1_MyWebApp>("mywebapp1");
builder.AddProject<Projects.TestingAppHost1_MyWorker>("myworker1")
    .WithEndpoint(name: "myendpoint1");
builder.AddPostgres("postgres1");
builder.Build().Run();
```

Then you can write tests which orchestrate and modify it like so: 
```csharp
[Fact]
public async Task CanRemoveResources()
{
    // Create a builder for the above AppHost - the type, Program, is any type in the app host's assembly.
    // Constructing the builder executes the AppHost on its own thread up until the point that its `Build()`
    // is called. At that point, the AppHost is suspended until our test builder's Build() is called, giving
    // the developer the opportunity to modify the app model before build.
    var appHost = new DistributedApplicationTestingBuilder<Program>();

    // Remove the redis resource from the app host - eg because we don't need it for this test.
    appHost.Resources.Remove(appHost.Resources.Single(r => r.Name == "redis1"));

    // Calling Build() allows the AppHost to resume building, returning the app.
    // The AppHost is suspended again it calls RunAsync or StartAsync. It will be resumed once
    // StartAsync is called in test code. When the app is disposed, the AppHost is terminated.
    await using var app = appHost.Build();
 
    // Allow the AppHost to start and wait for it to complete starting.
    await app.StartAsync();

    // Ensure that the resource which we added is present in the model.
    var appModel = app.Services.GetRequiredService<DistributedApplicationModel>();
    Assert.DoesNotContain(appModel.GetContainerResources(), c => c.Name == "redis1");
    Assert.Contains(appModel.GetProjectResources(), p => p.Name == "myworker1");

    // Get an endpoint from a resource
    var workerEndpoint = app.GetEndpoint("myworker1", "myendpoint1");
    Assert.NotNull(workerEndpoint);
    Assert.True(workerEndpoint.Host.Length > 0);

    // Get a connection string from a resource
    var pgConnectionString = app.GetConnectionString("postgres1");
    Assert.NotNull(pgConnectionString);
    Assert.True(pgConnectionString.Length > 0);

  // As the 'app' gets disposed, the AppHost is terminated and the resources it created are deleted.
}
```

For this to work, you currently need to add this snippet to your app host project (eg, at the bottom of Program.cs):

```csharp
public partial class Program
{
}
```

* `DistributedApplicationTestingHarness<TEntryPoint>`

This type is akin to `WebApplicationFactory<TEntryPoint>`, but for `DistributedApplicationBuilder` instead of host builders.

Here is an example which adds an xUnit fixture which uses the testing harness:

```csharp
public class TestingHarnessTests(DistributedApplicationFixture<Program> appHostFixture) : IClassFixture<DistributedApplicationFixture<Program>>
{
    private readonly HttpClient _httpClient = appHostFixture.CreateHttpClient("mywebapp1");

    [Fact]
    public async Task HttpClientGetTest()
    {
        var result1 = await _httpClient.GetFromJsonAsync<WeatherForecast[]>("/weatherforecast");
        Assert.NotNull(result1);
        Assert.True(result1.Length > 0);
    }

    private sealed record WeatherForecast(DateOnly Date, int TemperatureC, string? Summary)
    {
        public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
    }
}
```

The fixture:

```csharp
public sealed class DistributedApplicationFixture<TEntryPoint> : DistributedApplicationTestingHarness<TEntryPoint>, IAsyncLifetime where TEntryPoint : class
{
    protected override void OnBuilderCreating(DistributedApplicationOptions applicationOptions, HostApplicationBuilderSettings hostOptions)
    {
        // This does nothing, but shows what can be intercepted.
        base.OnBuilderCreating(applicationOptions, hostOptions);
    }

    protected override void OnBuilderCreated(DistributedApplicationBuilder applicationBuilder)
    {
        // This does nothing, but shows what can be intercepted.
        base.OnBuilderCreated(applicationBuilder);
    }

    protected override void OnBuilding(DistributedApplicationBuilder applicationBuilder)
    {
        // This does nothing, but shows what can be intercepted.
        base.OnBuilding(applicationBuilder);
    }

    public async Task InitializeAsync() => await base.InitializeAsync();

    async Task IAsyncLifetime.DisposeAsync() => await DisposeAsync();
}
```

The harness base class offers members to:
* Get the `DistributedApplication` via the `DistributedApplication` property
* Get an `HttpClient`, endpoint, or connection string for a resource

TODO:
* [ ] Pipe console output from orchestrated resources & app host to test output
* [ ] VS support for debugging the orchestrated resources

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2310)

### CODE REVIEW

**Comment by davidfowl** [comment]
@ReubenBond do you think there's anything incremental thats worth getting in for p4?

**Comment by ReubenBond** [comment]
Yes, it's good to have the basics available in p4. It has caveats like: The app projects won't be launched under the debugger, and ports aren't randomized, so your tests need to be part of the same xunit test collection to make them run sequentially

**Comment by davidfowl** [comment]
Moving to P5

**Comment by ReubenBond** [comment]
@DamianEdwards please take a look

File: src/Aspire.Hosting.Testing/DistributedApplicationEntryPointInvoker.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                _disposable = listener.Subscribe(this);
+            }
+
+            public void OnCompleted()
+            {
+                _disposable?.Dispose();
+            }
+
+            public void OnError(Exception error)
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Testing/DistributedApplicationEntryPointInvoker.cs:173]: ???

File: src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        if (!resource.TryGetAllocatedEndPoints(out var endpoints))
+        {
+            throw new InvalidOperationException($"Cannot create a client for resource '{resourceName}' because it has no allocated endpoints.");
+        }
+
+        AllocatedEndpointAnnotation? endpoint = null;
+
+        if (!string.IsNullOrEmpty(endpointName))
+        {
+            endpoint = endpoints.FirstOrDefault(e => string.Equals(e.Name, endpointName, StringComparison.OrdinalIgnoreCase));
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs:41]: SingleOrDefault. If we have an endpoint with the same name then we should consider it an exception.

File: src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    /// <summary>
+    /// Gets the connection string for the specified resource.
+    /// </summary>
+    /// <param name="app">The application.</param>
+    /// <param name="resourceName">The resource name.</param>
+    /// <returns>The connection string for the specified resource.</returns>
+    /// <exception cref="ArgumentException">The resource was not found or does not expose a connection string.</exception>
+    public static string? GetConnectionString(this DistributedApplication app, string resourceName)
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs:74]: @davidfowl awareness ... this will break with your incoming async changes.

**Comment by mitchdenny** [comment]
This is really good comprehensive stuff. Impressed with the approach of using the DiagnosticSource/Listener to extract internal state from the entry point. It is a shame that we need to use the partial class trick to coerce the right accessibility on the entry point.

BTW, I am not sure that resource removal is a great example. In your sample you `AddRedis(...)` but you don't reference it anywhere. If you did reference it you would find that the `WithReference(...)` call would have resulted in a callback which captures the `RedisResource`.

Your code would result in DCP not spinning up the container, but the callbacks would likely blow up.

File: src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs (hunk 2 above)
Comment (ReubenBond) [reply src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs:41]: We should catch that during app model build / launch, ideally. I'll add a safeguard here anyway

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting.Testing/DistributedApplicationTestingBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    { }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DistributedApplicationTestingBuilder{TEntryPoint}"/> class.
+    /// </summary>
+    /// <param name="configureBuilder">The delegate used to configure the creation of the builder.</param>
+    public DistributedApplicationTestingBuilder(Action<DistributedApplicationOptions, HostApplicationBuilderSettings> configureBuilder)
+    {
+        _factory = new(configureBuilder);
+        _applicationBuilder = _factory.DistributedApplicationBuilder.Result;
```
Comment (BrennanConroy) [inline src/Aspire.Hosting.Testing/DistributedApplicationTestingBuilder.cs:34]: This just hang forever. Are the tests missing using the fixture or something?

Comment (ReubenBond) [reply src/Aspire.Hosting.Testing/DistributedApplicationTestingBuilder.cs:34]: I introduced this issue when trying to make CI tests work. There is a fix PR: #2575

**Pull request merged**